scheduler/src/cook/scheduler/scheduler.clj

```diff
@@ -1232,16 +1238,43 @@
                     (launch-matched-tasks! matches conn db fenzo mesos-run-as-user pool-name)
                     (update-host-reservations! rebalancer-reservation-atom matched-job-uuids)
                     matched-considerable-jobs-head?))
+                ; Absolute maximum jobs we will consider autoscaling to.
+                {:keys [max-jobs-for-autoscaling autoscaling-scale-factor]} (config/kubernetes)
+                ; The fraction of jobs we tried to match that actually matched.
+                fraction-unmatched (if (pos? number-total) (/ (float number-unmatched) number-total) 0)
+                ; We want to autoscale any unmatched job, or
+                ; We want to scale our max-jobs-for-autoscaling by the fraction of the jobs we just matched.
+                ; E.g. If we matched 20% of the queue, then we want to autoscale to 20% of max-jobs-for-autoscaling)
+                ; We include a scale factor however, so that if we match 20$ and scale factor is 2.5, we'll generate
+                ; pods for 50% of max-jobs-for-autoscaling.
+                number-for-autoscaling (-> fraction-unmatched
+                                           (* autoscaling-scale-factor)
+                                           (min 1) ; Can't match more than 100% of max-jobs-for-autoscaling.
+                                           (* max-jobs-for-autoscaling)
+                                           int
+                                           (max number-unmatched)) ; Autoscale at least the pods that failed to match.
                 ;; We need to filter pending jobs based on quota so that we don't
                 ;; trigger autoscaling beyond what users have quota to actually run
                 autoscalable-jobs (->> pool-name
                                        (get @pool-name->pending-jobs-atom)
                                        (tools/filter-pending-jobs-for-quota pool-name (atom {}) (atom {})
-                                         user->quota user->usage (tools/global-pool-quota (config/pool-quotas) pool-name)))]
+                                                                            user->quota user->usage (tools/global-pool-quota (config/pool-quotas) pool-name))
+                                       (take number-for-autoscaling))
+                filtered-autoscalable-jobs (remove #(.getIfPresent caches/autoscale-retry-blacklist (:job/uuid %)) autoscalable-jobs)]
```

Comment:
What's the rationale for scaling first and then filtering? 

Seems to me like we'd want to filter first (lazily), then scale based on `fraction-unmatched`.

Reply:
It would create the wrong answer in that we could no longer tune the loop from first principals. This scaling loop was created so we could tone it from first principals. 

1.  outstanding synthetic pods is a function of the desired scaling rate which we choose, as well as the time to create a new pod (from GKE/k8s) This defines the peak rate we intend to support in terms of SLO. Note that this is just a number. (= scale rate / typ P95 synthetic pod launch time  / typ pod size)
2. We adjust this based on fraction matched. If we're matching heavily, we don't need to autoscale as aggressively. Note that this is a number, not a list. This defines our effective scaling demand. Scaling more than this means we're scaling above our desired scaling rate. Scaling under this means that we're scaling under our desired scaling rate. (fudge factor = 1.0 or 1.50) We make sure to always synthetic pod for any unmatched job, in case it has unusual machine requirements.
3. We take the list of jobs do hard filter for out of quota. don't pass any jobs that are out of quota downstream.
4. Next, we limit the jobs for synthetic pods to be our effective scaling demand. This, combined with the time to create a new synthetic pod means we're neither oversatisfying or undersatisfying our demand to autoscale.
5. We then have the filter as an optimization so that we don't pointlessly relaunch the same pod too frequently as a synthetic pod (when we're launch starved). When we're intentionally scaling, the synthetic pod launch time is the same duration as this filter block time and this filter is effectively a noop. It only comes into play when we're in a launch starved situation and, in some sense, emulates synthetic pods as having a longer time-to-complete, as seen by the rest of the scale-up system. This gives us the benefits of the optimization, without altering the scaling loop or scaling demand behavior.

We could do steps in the order 3, 1, 2, which may seem more clear, but as steps 1 and 2 only produce a number, doing them first doesn't change the result.

Doing step 5 before step 4 means that this design isn't constructed from first principals, so will have behavior that can be dysfunctional. In particular, #1, it means that we're potentially scaling over our target demand rate. #2 It can also dive 10x deeply into the queue, requesting synthetic pods for jobs that we didn't intend to have them requested this early, e.g., if they have one-off machine type requirements.

Reply:
Let's say Cook does the math for scaling based on `fraction-unmatched`, and the answer is that we should send 100 synthetic pods this cycle. And let's say the top 100 jobs in the pending queue all had a synthetic pod sent in the last, say, 1 minute. 

The current ordering of code will mean that we send 0 synthetic pods this cycle, and this situation could persist for a while (potentially minutes, depending on what we configure `recency-seconds` to), effectively halting all autoscaling during this time. This would be true even if we have thousands of pending jobs in the queue.

I don't think we want this behavior. In this example, I think we instead want to send synthetic pods for the top 100 jobs in the queue that have not recently had a synthetic pod launched for them.

Reply:
@scrosby Thoughts on this?

Reply:
Agree with the first paragraph, but now lets break this situation down further:

1. Lets say those 100 synthetic pods are needed in order to launch the workload, (we're at capacity or they have a unique machine type)  that means that these synthetic pods will continue to be outstanding until resources are created to run them --- I.e., we are actively autoscaling to launch the workload.  In this case, even if the filter wasn't there, we wouldn't be launching more pods --- we still have these 100 pods outstanding, and the only-one-synthetic-pod-per-job would suppress autoscaling further. So, in this case the filter is a noop; we're filtering jobs that would already be filtered downstream, doing the right thing, with higher performance.
2. Now, lets say those 100 synthetic pods are for a workload we can run now. however, we didn't match them to actually launch because they were beyond the considerable queue but still in the autoscaling queue. In this case, there's no point in creating synthetic pods for them every cycle, which the k8s scheduler will place and they'll complete. In this case, because these jobs don't require autoscaling resources to fit in, they're not evidence that we need to look even *further* into the queue for jobs. Again, the filter is doing the right thing, its avoiding k8s launch bandwidth.

What swapping does do is it makes an interesting functionality change. 
If we swap the order, we're basically avoiding the performance optimization this code intended and we're choosing to autoscale more aggresively, and in a strange way. In the first scenario above, there will be no change.  However, in the second scenario above, instead of wasting launches on the same synthetic pod in the above scenario, we're using that launch bandwidth to march another 100 pods in the queue each cycle for autoscaling because the filter suppreses that queue prefix. 

To be concrete, here's another scenario that illustrates the difference very concretely:
* The first 400 jobs in the queue are stuck waiting for a special nodetype to be created, and will be stuck unmatched for the next 2 minutes.
* The next 25k jobs can launch right now, however, no more than 1000 simultaneously.
* Our optimizations successfully bring the match cycle down to 5 seconds.

A. If we do no filter, then on each match cycle, we'll match no jobs, causing us to create synthetic pods for the next 1000 jobs, they'll launch as we have resources, and we'll try repeatedly launching those next 1000 jobs 24 times over the next 120 seconds.
B. If we do the take, then filter, then on each match cycle, we'll match no jobs, causing us to want to create synthetic pods for the next 1000 jobs. The first iteration through, we'll create them, they'll launch immediately and then complete. For the next 23 iterations, the filter will suppress them.
C. If we filter, then take, then on each match cycle, we'll match no jobs, causing us to create synthetic pods for the next 1000 jobs, they'll launch and complete. The next cycle, the filter skips jobs 400-1400 in the queue, so we'll launch jobs #1400-2400 in the queue as synthetic pods. We'll repeat 23 times and launch synthetic pods for next 24000 jobs in the queue over the next 120 seconds.

The same principal applies even if the entire queue isn't stuck for 120 seconds. 

So, my thoughts on this functionality change embodied in C:
- C is slower than B, going contrary to the goal of optimizing our code.
- C launches just as many synthetic pods as A, and much more than B. 
- I expect C to be somewhat slower than A; it will visit much more of the queue and that has implications for cache hit rate and laziness.
- C will autoscale more aggressively than A, this can be mixed. We may waste resources more aggressively autoscaling, but may also decrease autoscaling latency for resources. However, personally, if we wanted to increase autoscaling aggressiveness, we could double max-jobs-for-autoscaling at a lot less cost.

Thus,. my vote for B.

Reply:
makes sense to me, thanks! 👍🏼